### PR TITLE
cxx-rs: init at 1.0.72

### PR DESCRIPTION
**Pull request by centromere**
###### Description of changes

Add [cxx](https://github.com/dtolnay/cxx). This is a necessary [dependency](https://github.com/zcash/zcash/blob/882935e35ee338876c071f9cd93aa3694154de73/depends/packages/native_cxxbridge.mk) of zcash 5.1.0.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/cxx-rs/default.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  version = "1.0.72";
+
+  src = fetchFromGitHub {
+    owner = "dtolnay";
+    repo = "cxx";
+    rev = version;
+    sha256 = "sha256-+OumeLSgz8kLQKhEc3icCk1q+X+S7Xt+XtAlkx8iguU=";
+  };
+
+  cargoPatches = [
```
Comment (K900) [inline pkgs/development/libraries/cxx-rs/default.nix:14]: If you're doing this, might as well just put the Cargo.lock in-tree and remove cargoSha256.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -35963,4 +35963,6 @@ with pkgs;
   swift-corelibs-libdispatch = callPackage ../development/libraries/swift-corelibs-libdispatch { };
 
   swaysettings = callPackage ../applications/misc/swaysettings { };
+
+  cxx-rs = callPackage ../development/libraries/cxx-rs { };
```
Comment (K900) [inline pkgs/top-level/all-packages.nix:35967]: The list is somewhat alphabetical, maybe try to keep it that way.

File: pkgs/development/libraries/cxx-rs/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  cargoTestFlags = [ "--workspace" ];
+
+  outputs = [ "out" "doc" "dev" ];
+
+  postInstall = ''
+    mkdir -p $doc
+    cp -r ./target/doc/* $doc
+
+    mkdir -p $dev/include/rust
+    install -D -m 0644 ./include/cxx.h $dev/include/rust
```
[the remark this answers is not in the document]
Comment (centromere) [reply pkgs/development/libraries/cxx-rs/default.nix:40]: Various [examples](https://cxx.rs/tutorial.html?highlight=rust%2Fcxx.#shared-data-structures) in the documentation refer to that path:

Screen Shot 2022-07-21 at 11 31 54

I am open to suggestions for a different path, as I don't know of a more appropriate one.

File: pkgs/development/libraries/cxx-rs/default.nix [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  src = fetchFromGitHub {
+    owner = "dtolnay";
+    repo = "cxx";
+    rev = version;
+    sha256 = "sha256-+OumeLSgz8kLQKhEc3icCk1q+X+S7Xt+XtAlkx8iguU=";
+  };
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/cxx-rs/default.nix:16]: Can we avoid the lockfile by using fetchCrate?

File: pkgs/development/libraries/cxx-rs/default.nix [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    package = cxx-rs;
+    command = "cxxbridge --version";
+  };
+
+  meta = with lib; {
+    description = "Safe FFI between Rust and C++";
+    homepage = "https://github.com/dtolnay/cxx";
+    license = licenses.mit;
+    maintainers = with maintainers; [ centromere ];
+    platforms = platforms.linux ++ platforms.darwin;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/cxx-rs/default.nix:53]:
```suggestion
```
that is the default

Comment (centromere) [reply pkgs/development/libraries/cxx-rs/default.nix:53]: Thank you.

File: pkgs/development/libraries/cxx-rs/default.nix (hunk 4 above)
Comment (centromere) [reply pkgs/development/libraries/cxx-rs/default.nix:16]: When using `fetchCrate` the tests fail [and the author has declined to fix them](https://github.com/dtolnay/cxx/issues/1071).

However, the tests do pass when the source is checked out from git. While I can patch `Cargo.toml` in this way:

```
[patch.crates-io]
cxx-test-suite = { git = "https://github.com/dtolnay/cxx.git" }
```

it leads to a conflict in the dependency tree:

```
├── cxx-test-suite v0.0.0 (https://github.com/dtolnay/cxx.git#9e0d3680)
│   └── cxx v1.0.72 (https://github.com/dtolnay/cxx.git#9e0d3680)
│       ├── cxxbridge-macro v1.0.72 (proc-macro) (https://github.com/dtolnay/cxx.git#9e0d3680)
```

`cxx-test-suite` is being compiled against the version of `cxx` which exists in git, not the version in the current working directory. I have not figured out a way to address this. Any ideas?

**Pull request merged**